src/main/scala/mesosphere/marathon/core/task/Task.scala

```diff
@@ -194,98 +315,77 @@ object Task {
-        case _ => throw new MatchError(s"taskId $taskId is no valid identifier")
+    /**
+      * Parse instance and task id from idString.
+      *
+      * The string has to match one of
+      *   * LegacyTaskIdRegex
+      *   * ResidentTaskIdRegex
+      *   * TaskIdWithInstanceIdRegex
+      *   * ResidentTaskIdWithInstanceIdRegex
+      *
+      * @param idString The raw id that should be parsed.
+      * @return Task.Id
+      */
+    def fromIdString(idString: String): Task.Id = {
+      idString match {
+        case ResidentTaskIdWithInstanceIdRegex(safeRunSpecId, prefix, uuid, container, attempt) =>
+          val runSpec = PathId.fromSafePath(safeRunSpecId)
+          val instanceId = Instance.Id(runSpec, Instance.Prefix.fromString(prefix), UUID.fromString(uuid))
+          val containerName: Option[String] = if (container == Names.anonymousContainer) None else Some(container)
+          ResidentTaskId(instanceId, containerName, attempt.toLong)
+        case TaskIdWithInstanceIdRegex(safeRunSpecId, prefix, uuid, container) =>
+          val runSpec = PathId.fromSafePath(safeRunSpecId)
+          val instanceId = Instance.Id(runSpec, Instance.Prefix.fromString(prefix), UUID.fromString(uuid))
+          val containerName: Option[String] = if (container == Names.anonymousContainer) None else Some(container)
+
+          // This could also be a ResidentTaskId
+          EphemeralOrReservedTaskId(instanceId, containerName)
+        case ResidentTaskIdRegex(safeRunSpecId, separator, uuid, _, attempt) =>
+          val runSpec = PathId.fromSafePath(safeRunSpecId)
+          LegacyResidentId(runSpec, separator, UUID.fromString(uuid), attempt.toLong)
+        case LegacyTaskIdRegex(safeRunSpecId, separator, uuid) =>
+          val runSpec = PathId.fromSafePath(safeRunSpecId)
+          LegacyId(runSpec, separator, UUID.fromString(uuid))
+        case _ => throw new MatchError(s"taskId $idString no valid identifier")
       }
     }
 
-    def apply(mesosTaskId: MesosProtos.TaskID): Id = new Id(mesosTaskId.getValue)
+    /**
+      * Construct a task id from a Mesos task id.
+      *
+      * @param mesosTaskId The task identifier in the Mesos world.
+      * @return Task id in Marathon world.
+      */
+    def apply(mesosTaskId: MesosProtos.TaskID): Id = fromIdString(mesosTaskId.getValue)
 
     /**
-      * Create a taskId according to the old schema (no instance designator, no mesos container name).
+      * Create a taskId according to the old schema (no instance designator, no Mesos container name).
       * Use this when needing to create an ID for a normal App task or a task for initial reservation handling.
       *
       * Use @forResidentTask when you want to launch a task on an existing reservation.
       */
-    def forRunSpec(id: PathId): Id = {
-      val taskId = id.safePath + "." + uuidGenerator.generate()
-      Task.Id(taskId)
-    }
+    @deprecated("Task ids should be created from instance ids and not run spec ids")
```

Comment:
Why don't we just remove this method?

Reply:
I have another diff for this. If I'd remove it here the PR would blow up.